packages/code-snippet/src/CodeSnippetDisplay.tsx

```diff
@@ -0,0 +1,263 @@
+    const widget: Widget = this.props.getCurrentWidget();
+    const snippetStr: string = snippet.code.join('\n');
+
+    if (
+      widget instanceof DocumentWidget &&
+      (widget as DocumentWidget).content instanceof FileEditor
+    ) {
+      const documentWidget = widget as DocumentWidget;
+      const fileEditor = (documentWidget.content as FileEditor).editor;
+      const markdownRegex = /^\.(md|mkdn?|mdown|markdown)$/;
+      if (PathExt.extname(widget.context.path).match(markdownRegex) !== null) {
+        // Wrap snippet into a code block when inserting it into a markdown file
+        fileEditor.replaceSelection(
+          '```' + snippet.language + '\n' + snippetStr + '\n```'
+        );
+      } else if (widget.constructor.name == 'PythonFileEditor') {
+        this.verifyLanguageAndInsert(snippet, 'python', fileEditor);
+      } else {
+        fileEditor.replaceSelection(snippetStr);
+      }
+    } else if (widget instanceof NotebookPanel) {
+      const notebookWidget = widget as NotebookPanel;
+      const notebookCell = (notebookWidget.content as Notebook).activeCell;
+      const notebookCellEditor = notebookCell.editor;
+
+      if (notebookCell instanceof CodeCell) {
+        const kernelInfo = await notebookWidget.sessionContext.session?.kernel
+          ?.info;
+        const kernelLanguage: string = kernelInfo?.language_info.name || '';
+        this.verifyLanguageAndInsert(
+          snippet,
+          kernelLanguage,
+          notebookCellEditor
+        );
+      } else if (notebookCell instanceof MarkdownCell) {
+        // Wrap snippet into a code block when inserting it into a markdown cell
+        notebookCellEditor.replaceSelection(
+          '```' + snippet.language + '\n' + snippetStr + '\n```'
+        );
+      } else {
+        notebookCellEditor.replaceSelection(snippetStr);
+      }
+    } else {
+      this.showErrDialog('Code snippet insert failed: Unsupported widget');
+    }
+  };
+
+  // Handle language compatibility between code snippet and editor
+  private verifyLanguageAndInsert = async (
+    snippet: ICodeSnippet,
+    editorLanguage: string,
+    editor: CodeEditor.IEditor
+  ): Promise<void> => {
+    const snippetStr: string = snippet.code.join('\n');
+    if (
+      editorLanguage &&
+      snippet.language.toLowerCase() !== editorLanguage.toLowerCase()
+    ) {
+      const result = await this.showWarnDialog(
+        editorLanguage,
+        snippet.displayName
+      );
+      if (result.button.accept) {
+        editor.replaceSelection(snippetStr);
+      }
+    } else {
+      // Language match or editorLanguage is unavailable
+      editor.replaceSelection(snippetStr);
+    }
+  };
+
+  // Display warning dialog when inserting a code snippet incompatible with editor's language
+  private showWarnDialog = async (
+    editorLanguage: string,
+    snippetName: string
+  ): Promise<Dialog.IResult<string>> => {
+    return showDialog({
+      title: 'Warning',
+      body:
```

Comment:
this may be easier to read using template literals, i.e.,
```
body: `Code snippet "${snippetName}" is incompatible with ${editorLanguage}. Continue?`
```